File: 647.回文子串.java
/*
 * @lc app=leetcode.cn id=647 lang=java
 *
 * [647] 回文子串
 */

// @lc code=start
class Solution {
    public int countSubstrings(String s) {
        int len = s.length();
        boolean[][] dp = new boolean[len+1][len];
        for (int i = 0; i < len; i++) {
            dp[0][i] = true;
            dp[1][i] = true;
        }
        int res = 0;
        for (int i = 2; i < len+1; i++) {
            for (int j = 0; j <len-i+1; j++) {
                dp[i][j] = s.charAt(j) == s.charAt(i+j-1) && dp[i-2][j+1];
                if (dp[i][j])
                    res++;
            }
        }
        return res+len;
    }
}
// @lc code=end

File: 647.回文子串.py
#
# @lc app=leetcode.cn id=647 lang=python3
#
# [647] 回文子串
#

# @lc code=start


class Solution:
    def countSubstrings(self, s: str) -> int:
        s_len = len(s)
        dp = [[False for i in s] for i in range(s_len+1)]
        for i in range(s_len):
            dp[0][i] = True
            dp[1][i] = True
        res = 0
        for i in range(2, s_len+1):
            for j in range(s_len - i + 1):
                dp[i][j] = s[j] == s[j + i - 1] and dp[i - 2][j + 1]
                if dp[i][j]:
                    res += 1
        return res+s_len
# @lc code=end
